python/flare/code_grafter.py

```diff
@@ -0,0 +1,858 @@
+        """Add segments, inject stub functions, and patch calls within IDB to
+        point to the stubs.
+
+        Args:
+            mem (numbers.Integral): Size of segment to use for memory
+                allocators. The more generous, the less likelihood of running
+                out of memory due to successive calls to e.g. malloc and
+                crashing your emulator, Bochs debug session, etc. Unless or
+                until someone is brave enough to implement a stub malloc
+                implementation that supports free(), this unfortunately has to
+                be pretty sizeable in practice.
+        """
+        if self._findGraftedSegments():
+            msg = ('Found grafted code segments; use '
+                   'removeGraftedCodeFromIdb() first if you want to graft '
+                   'anew')
+            raise CodeGraftingAlreadyPresent(msg)
+        self._addSegments(mem)
+        return self._patchCalls()
+
+    def removeGraftedCodeFromIdb(self):
+        grafted = self._findGraftedSegments()
+        if grafted:
+            self._unpatchCalls(grafted)
+
+        for seg in grafted:
+            idc.del_segm(seg.start, idc.SEGMOD_KILL)
+
+    def _findGraftedSegments(self):
+        return [s for s in SegPlanner() if
+                idc.get_cmt(s.start, 1) == g_seg_sig_code_grafter]
+
+    def _addSegments(self, mem=0x4000000):
+        """Create emulation stub segments.
+
+        Includes generous memory allocation space by default.
+        """
+        arena_seg_size = mem
+        code_seg_size = 0x1000
+
+        use32 = 1
+        fmt_ptr_width = '<I'
+        if self.bits == 64:
+            use32 = 2
+            fmt_ptr_width = '<Q'
+
+        def le_hex(n):  # Little-endian N-bit hex
+            return binascii.hexlify(struct.pack(fmt_ptr_width, n))
+
+        seg_plan = SegPlanner()
+
+        # Pick locations for the code and the malloc "arena"
+        #
+        # Note 1: Find space for stub code segment first, making use of
+        # SegPlanner's bias in planning new segments close to other segments if
+        # possible. This violates the Franklin-Covey principle of fitting big
+        # stones in first before smaller stones create spatial fragmentation in
+        # the jar. But in a 64-bit IDB, depending on the size of the malloc
+        # arena, doing this in the opposite order could increase the chance of
+        # a stub function residing at a distance from its callsite that cannot
+        # be represented in 32 bits.
+        #
+        # Note 2: SegPlanner ensures segments won't start at 0, which otherwise
+        # could result in a NULL return from an allocator stub like malloc
+        # erroneously signifying failure.
+        code = seg_plan.addSegAnywhere(code_seg_size)
+        arena = seg_plan.addSegAnywhere(arena_seg_size)
+
+        for seg in (code, arena):
+            idc.AddSeg(seg.start, seg.end, 0, use32, 0, idc.scPub)
```

Comment:
gaaaah. i love ida's re-vamped & clean API. can't believe AddSeg is still a thing.

Reply:
Yeah 💯 feels like a lot of busywork to migrate to it since it's not like the APIs conform to any uniform nomenclature on this side of version 7.x either.

I previously looked at what you cleverly referred to as your IDA 7.0 Rosetta Stone (local copy of `idc_bc695.py`) to make sure this was still the right name to use. Since you mentioned it , I double checked in one additional place that I found useful for this:

<https://www.hex-rays.com/products/ida/support/ida74_idapython_no_bc695_porting_guide/>

It still prescribes `AddSeg` :-)